code/game/objects/structures/crematorium.dm

```diff
@@ -0,0 +1,254 @@
+								shake_animation(3)
+								playsound(loc, 'sound/effects/grillehit.ogg', 45, 1)
+							if(3)
+								playsound(src, 'sound/effects/bang.ogg', 45, 1)
+								if(prob(50))
+									playsound(src, 'sound/effects/bang.ogg', 45, 1)
+									shake_animation()
+								else
+									shake_animation(5)
+							if(4)
+								playsound(src, 'sound/effects/clang.ogg', 45, 1)
+								shake_animation(5)
+							if(5)
+								playsound(src, 'sound/weapons/smash.ogg', 50, 1)
+								if(prob(50))
+									playsound(src, 'sound/weapons/smash.ogg', 50, 1)
+									shake_animation(9)
+								else
+									shake_animation()
+
+			if(round_is_spooky())
+				if(prob(50))
+					playsound(src, 'sound/effects/ghost.ogg', 10, 5)
+				else
+					playsound(src, 'sound/effects/ghost2.ogg', 10, 5)
+
+			if (!M.stat)
+				M.audible_message("[M]'s screams cease, as does any movement within the [src]. All that remains is a dull, empty silence.")
+
+			admin_attack_log(M, A, "Cremated their victim.", "Was cremated.", "cremated")
+			M.dust()
+
+		for(var/obj/O in contents) //obj instead of obj/item so that bodybags and ashes get destroyed. We dont want tons and tons of ash piling up
+			if(!istype(O, connected_tray))
+				qdel(O)
+
+		new /obj/effect/decal/cleanable/ash(src)
+		sleep(30)
+		cremating = initial(cremating)
+		locked = initial(locked)
+		playsound(src, 'sound/effects/spray.ogg', 50, 1)
+		update_icon()
+
+/obj/structure/crematorium_tray
+	name = "crematorium tray"
+	desc = "Apply body before burning."
+	icon = 'icons/obj/structures/crematorium.dmi'
+	icon_state = "crematorium_tray"
+	density = TRUE
+	anchored = TRUE
+	throwpass = TRUE
+	layer = BELOW_OBJ_LAYER
+
+	var/obj/structure/crematorium/connected_crematorium
+
+/obj/structure/crematorium_tray/Destroy()
+	if(!QDELETED(connected_crematorium))
+		QDEL_NULL(connected_crematorium)
+	return ..()
+
+/obj/structure/crematorium_tray/attack_hand(mob/user)
+	if(Adjacent(user))
+		connected_crematorium.attack_hand(user)
+	return ..()
+
+/obj/structure/crematorium_tray/attack_robot(mob/user)
+	if(Adjacent(user))
+		attack_hand(user)
+
+/obj/structure/crematorium_tray/receive_mouse_drop(atom/dropping, mob/user)
+	. = ..()
+	if(!. && (ismob(dropping) || istype(dropping, /obj/structure/closet/body_bag)))
+		var/atom/movable/AM = dropping
+		if(!AM.anchored)
+			AM.forceMove(loc)
+			if(user != dropping)
+				user.visible_message(SPAN_NOTICE("\The [user] stuffs \the [dropping] onto \the [src]!"))
+			return TRUE
+
+/obj/machinery/button/crematorium
+	name = "crematorium igniter"
+	desc = "Burn baby burn!"
+	icon = 'icons/obj/power.dmi'
+	icon_state = "crematorium_switch"
+	initial_access = list(access_crematorium)
+
+/obj/machinery/button/crematorium/on_update_icon()
+	return
+
+/obj/machinery/button/crematorium/activate(mob/user)
+	if(operating)
+		return
+
+	for(var/obj/structure/crematorium/C in range())
+		if (C.id_tag == id_tag)
+			if (!C.cremating)
+				C.cremate(user)
```

Comment:
I think you should have it activate all null-id crematoriums, maybe, or just all crematoriums in range (how many can there be? No, don't answer that...). You can't set the id on the crematorium but can on the button, which makes this interaction weird.